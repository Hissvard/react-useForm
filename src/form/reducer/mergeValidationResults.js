import { Map, List } from 'immutable'

const getError = (errors, key) => {
  const error = errors[key]

  if (typeof error === 'string') return error
  if (Array.isArray(error) && error.error) {
    return error.error
  }

  return ''
}

const updateItems = (list, errorObj) => {
  const errors = errorObj || {}

  return list.map((listItem, index) => {
    const error = Array.isArray(errors) ? errors[index] : errors
    return listItem.update('fields', Map(), fields => updateMap(fields, error))
  })
}

const updateMap = (fields, errorObj) => {
  const errors = errorObj || {}

  return fields.entrySeq().reduce((acc, [key, field]) => {
    const next = field.get('items') ?
      field.update('items', List(), items => updateItems(items, errors[key])) :
      field


    const error = getError(errors, key)

    return acc
      .mergeIn([key], next)
      .setIn([key, 'current', 'error'], Boolean(error))
      .setIn([key, 'current', 'helperText'], error)

  }, Map())

}

export const mergeValidationResults = (state, errors = {}) => {
  return state.update('fields', Map(), fields => updateMap(fields, errors))
}
