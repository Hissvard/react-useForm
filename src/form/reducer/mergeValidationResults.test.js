import { fromJS } from 'immutable'
import { mergeValidationResults } from './mergeValidationResults'

let state
beforeEach(() => {
  state = fromJS({
    other: 'other data',
    fields: {
      nickname: {
        current: {
          error: 'notSet',
          helperText: 'notSet',
        },
        initial: 'initialNickname',
      },
      fullName: {
        current: {
          error: 'notSet',
          helperText: 'notSet',
        },
      },
      colors: {
        current: {
          error: 'notSet',
          helperText: 'notSet',
        },
        initial: 'initialColors',
        items: [
          {
            fields: {
              color: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
                initial: 'initialColor',
              },
            },
          },
          {
            fields: {
              color: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
              },
              shades: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
                items: [
                  {
                    fields: {
                      shade: {
                        current: {
                          error: 'notSet',
                          helperText: 'notSet',
                        },
                        initial: 'initialShade',
                      },
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  })
})

test('no errors will clear state values', () => {
  expect(mergeValidationResults(state, undefined)).toMatchSnapshot()
})

test('sets a single error', () => {
  const errors = { fullName: 'i am not full' }
  expect(mergeValidationResults(state, errors)).toMatchSnapshot()
})

test('sets an error on a list', () => {
  const errors = { colors: 'blind' }
  expect(mergeValidationResults(state, errors)).toMatchSnapshot()
})

test('sets an error on a list field', () => {
  const errors = {
    colors: [
      null,
      { color: 'blind' },
    ],
  }
  expect(mergeValidationResults(state, errors)).toMatchSnapshot()
})

test('sets an error on a list and its field', () => {
  const errors = {
    colors: [
      null,
      { color: 'blind' },
    ],
  }
  errors.colors.error = 'no blindness'

  expect(mergeValidationResults(state, errors)).toMatchSnapshot()
})

test('sets an error on a nested list field', () => {
  const errors = {
    colors: [
      { color: 'blind' },
      {
        shades: [
          { shade: 'down' },
        ],
      },
    ],
  }
  expect(mergeValidationResults(state, errors)).toMatchSnapshot()
})
