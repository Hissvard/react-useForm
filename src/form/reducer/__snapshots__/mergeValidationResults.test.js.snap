// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no errors will clear state values 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": false,
                        "helperText": "",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`sets a single error 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": true,
        "helperText": "i am not full",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": false,
                        "helperText": "",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`sets an error on a list 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": true,
        "helperText": "blind",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": false,
                        "helperText": "",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`sets an error on a list and its field 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": true,
        "helperText": "no blindness",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": true,
                "helperText": "blind",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": false,
                        "helperText": "",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`sets an error on a list field 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": true,
                "helperText": "blind",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": false,
                        "helperText": "",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`sets an error on a nested list field 1`] = `
Immutable.Map {
  "other": "other data",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialNickname",
    },
    "fullName": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": false,
        "helperText": "",
      },
      "initial": "initialColors",
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": true,
                "helperText": "blind",
              },
              "initial": "initialColor",
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": false,
                "helperText": "",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": true,
                        "helperText": "down",
                      },
                      "initial": "initialShade",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;
